### PR TITLE
Add a simple `nbytes` representation in DataArrays and Dataset `repr` (opt-in)

### LINKED ISSUE #8690: Add `nbytes` to repr?
### Is your feature request related to a problem?

Would having the `nbytes` value in the `Dataset` repr be reasonable?

I frequently find myself logging this separately. For example:

```diff
<xarray.Dataset>
Dimensions:  (lat: 25, time: 2920, lon: 53)
Coordinates:
  * lat      (lat) float32 75.0 72.5 70.0 67.5 65.0 ... 25.0 22.5 20.0 17.5 15.0
  * lon      (lon) float32 200.0 202.5 205.0 207.5 ... 322.5 325.0 327.5 330.0
  * time     (time) datetime64[ns] 2013-01-01 ... 2014-12-31T18:00:00
Data variables:
-    air      (time, lat, lon) float32 dask.array<chunksize=(2920, 25, 53), meta=np.ndarray>
+    air      (time, lat, lon) float32 15MB dask.array<chunksize=(2920, 25, 53), meta=np.ndarray> 
Attributes:
    Conventions:  COARDS
    title:        4x daily NMC reanalysis (1948)
    description:  Data is from NMC initialized reanalysis\n(4x/day).  These a...
    platform:     Model
    references:   http://www.esrl.noaa.gov/psd/data/gridded/data.ncep.reanaly...
```

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

Status quo :)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by etienneschalk**
Edit: in contrary to what the title suggest, this is not an opt-in feature, it is enabled by default 

- [x] Closes #8690
   - (or at least is a proposal)
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- ~~[ ] New functions/methods are listed in `api.rst`~~

### CODE REVIEW

File: xarray/tests/test_formatting.py [hunk 1]
```diff
@@ -818,3 +819,59 @@ def test_empty_cftimeindex_repr() -> None:
...
+array([   0,    1,    2, ..., 1197, 1198, 1199], dtype=int16)
+Coordinates:
+  * foo      (foo) int16 0 1 2 3 4 5 6 7 ... 1193 1194 1195 1196 1197 1198 1199
+""".strip()
+        assert actual == expected
+
+    with xr.set_options(display_nbytes=True):
+        actual = repr(xds)
+        expected = """
+<xarray.Dataset 3kB>
```
Comment (dcherian) [inline xarray/tests/test_formatting.py:860]: How about 
```suggestion
<xarray.Dataset> Size: 3kB
```

**Comment by dcherian** [comment]
I don't think we need yet another global option for the repr. This takes up very little extra space so we can just show it always.

**Comment by etienneschalk** [comment]
> Thanks @etienneschalk . Very nice! However I don't think we need yet another global option for the repr. This takes up very little extra space so we can just show it always.

Hello,
Thanks for the review!
I originally added the feature flag to not have to change existing tests, also for the feature to be an opt-in. After having enabled it, I noticed that many tests and doctests are failing due to the change in repr (looking for string `    <xarray.` in the code yields 800+ results.
Before making these many changes, maybe the new representation should be rediscussed and consolidated and validated by others (not to have to do these changes twice)

**Comment by etienneschalk** [comment]
Procedure used:

- As mentioned in the issue https://github.com/max-sixty/pytest-accept/issues/146, backslashes are not supported, so they are removed
- Then doctests is ran on multiple files:

```bash
pytest --doctest-modules xarray/backends/api.py --accept
pytest --doctest-modules xarray/backends/common.py --accept
pytest --doctest-modules xarray/backends/file_manager.py --accept
pytest --doctest-modules xarray/backends/h5netcdf_.py --accept
pytest --doctest-modules xarray/backends/locks.py --accept
pytest --doctest-modules xarray/backends/lru_cache.py --accept
pytest --doctest-modules xarray/backends/memory.py --accept
pytest --doctest-modules xarray/backends/netcdf3.py --accept
pytest --doctest-modules xarray/backends/netCDF4_.py --accept
pytest --doctest-modules xarray/backends/plugins.py --accept
pytest --doctest-modules xarray/backends/pydap_.py --accept
pytest --doctest-modules xarray/backends/pynio_.py --accept
pytest --doctest-modules xarray/backends/scipy_.py --accept
pytest --doctest-modules xarray/backends/store.py --accept
pytest --doctest-modules xarray/backends/zarr.py --accept
pytest --doctest-modules xarray/coding/calendar_ops.py --accept
pytest --doctest-modules xarray/coding/cftimeindex.py --accept
pytest --doctest-modules xarray/coding/cftime_offsets.py --accept
pytest --doctest-modules xarray/coding/frequencies.py --accept
pytest --doctest-modules xarray/coding/strings.py --accept
pytest --doctest-modules xarray/coding/times.py --accept
pytest --doctest-modules xarray/coding/variables.py --accept
pytest --doctest-modules xarray/conventions.py --accept
pytest --doctest-modules xarray/convert.py --accept
pytest --doctest-modules xarray/core/accessor_dt.py --accept
pytest --doctest-modules xarray/core/accessor_str.py --accept
pytest --doctest-modules xarray/core/_aggregations.py --accept
pytest --doctest-modules xarray/core/alignment.py --accept
pytest --doctest-modules xarray/core/arithmetic.py --accept
pytest --doctest-modules xarray/core/combine.py --accept
pytest --doctest-modules xarray/core/common.py --accept
pytest --doctest-modules xarray/core/computation.py --accept
pytest --doctest-modules xarray/core/concat.py --accept
pytest --doctest-modules xarray/core/coordinates.py --accept
pytest --doctest-modules xarray/core/dask_array_ops.py --accept
pytest --doctest-modules xarray/core/daskmanager.py --accept
pytest --doctest-modules xarray/core/dataarray.py --accept
pytest --doctest-modules xarray/core/dataset.py --accept
pytest --doctest-modules xarray/core/dtypes.py --accept
pytest --doctest-modules xarray/core/duck_array_ops.py --accept
pytest --doctest-modules xarray/core/extensions.py --accept
pytest --doctest-modules xarray/core/formatting_html.py --accept
pytest --doctest-modules xarray/core/formatting.py --accept
pytest --doctest-modules xarray/core/groupby.py --accept
pytest --doctest-modules xarray/core/indexes.py --accept
pytest --doctest-modules xarray/core/indexing.py --accept
pytest --doctest-modules xarray/core/merge.py --accept
pytest --doctest-modules xarray/core/missing.py --accept
pytest --doctest-modules xarray/core/nanops.py --accept
pytest --doctest-modules xarray/core/npcompat.py --accept
pytest --doctest-modules xarray/core/nputils.py --accept
pytest --doctest-modules xarray/core/ops.py --accept
pytest --doctest-modules xarray/core/options.py --accept
pytest --doctest-modules xarray/core/parallelcompat.py --accept
pytest --doctest-modules xarray/core/parallel.py --accept
pytest --doctest-modules xarray/core/pdcompat.py --accept
pytest --doctest-modules xarray/core/pycompat.py --accept
pytest --doctest-modules xarray/core/resample_cftime.py --accept
pytest --doctest-modules xarray/core/resample.py --accept
pytest --doctest-modules xarray/core/rolling_exp.py --accept
pytest --doctest-modules xarray/core/rolling.py --accept
pytest --doctest-modules xarray/core/_typed_ops.py --accept
pytest --doctest-modules xarray/core/types.py --accept
pytest --doctest-modules xarray/core/utils.py --accept
pytest --doctest-modules xarray/core/variable.py --accept
pytest --doctest-modules xarray/core/weighted.py --accept
pytest --doctest-modules xarray/datatree_/conftest.py --accept
pytest --doctest-modules xarray/datatree_/datatree/common.py --accept
pytest --doctest-modules xarray/datatree_/datatree/datatree.py --accept
pytest --doctest-modules xarray/datatree_/datatree/extensions.py --accept
pytest --doctest-modules xarray/datatree_/datatree/formatting_html.py --accept
pytest --doctest-modules xarray/datatree_/datatree/formatting.py --accept
pytest --doctest-modules xarray/datatree_/datatree/io.py --accept
pytest --doctest-modules xarray/datatree_/datatree/iterators.py --accept
pytest --doctest-modules xarray/datatree_/datatree/mapping.py --accept
pytest --doctest-modules xarray/datatree_/datatree/ops.py --accept
pytest --doctest-modules xarray/datatree_/datatree/render.py --accept
pytest --doctest-modules xarray/datatree_/datatree/testing.py --accept
pytest --doctest-modules xarray/datatree_/datatree/tests/conftest.py --accept
pytest --doctest-modules xarray/datatree_/datatree/tests/test_dataset_api.py --accept
pytest --doctest-modules xarray/datatree_/datatree/tests/test_datatree.py --accept
pytest --doctest-modules xarray/datatree_/datatree/tests/test_extensions.py --accept
pytest --doctest-modules xarray/datatree_/datatree/tests/test_formatting_html.py --accept
pytest --doctest-modules xarray/datatree_/datatree/tests/test_formatting.py --accept
pytest --doctest-modules xarray/datatree_/datatree/tests/test_io.py --accept
pytest --doctest-modules xarray/datatree_/datatree/tests/test_mapping.py --accept
pytest --doctest-modules xarray/datatree_/datatree/tests/test_treenode.py --accept
pytest --doctest-modules xarray/datatree_/datatree/tests/test_version.py --accept
pytest --doctest-modules xarray/datatree_/datatree/treenode.py --accept
pytest --doctest-modules xarray/datatree_/docs/source/conf.py --accept
pytest --doctest-modules xarray/namedarray/_aggregations.py --accept
pytest --doctest-modules xarray/namedarray/_array_api.py --accept
pytest --doctest-modules xarray/namedarray/core.py --accept
pytest --doctest-modules xarray/namedarray/dtypes.py --accept
pytest --doctest-modules xarray/namedarray/_typing.py --accept
pytest --doctest-modules xarray/namedarray/utils.py --accept
pytest --doctest-modules xarray/plot/accessor.py --accept
pytest --doctest-modules xarray/plot/dataarray_plot.py --accept
pytest --doctest-modules xarray/plot/dataset_plot.py --accept
pytest --doctest-modules xarray/plot/facetgrid.py --accept
pytest --doctest-modules xarray/plot/utils.py --accept
pytest --doctest-modules xarray/testing/assertions.py --accept
pytest --doctest-modules xarray/testing/strategies.py --accept
pytest --doctest-modules xarray/tutorial.py --accept
pytest --doctest-modules xarray/util/deprecation_helpers.py --accept
pytest --doctest-modules xarray/util/generate_aggregations.py --accept
pytest --doctest-modules xarray/util/generate_ops.py --accept
pytest --doctest-modules xarray/util/print_versions.py --accept
```

Finally the backslash-removing commit is reverted

**Comment by max-sixty** [comment]
Hi @etienneschalk — this looks really good — thank you!

The only issue I see is that the Windows tests seem to have different values for the size in the `repr` tests, [for example](https://github.com/pydata/xarray/actions/runs/7792432615/job/21250418463#step:9:1550):

```
  - <xarray.Dataset> Size: 1kB
  ?                        ^^
  + <xarray.Dataset> Size: 640B
  ?                        ^^^
```

Is anyone familiar with what's going on here?

I guess we could exclude those in windows if there isn't an easily reconcilable approach...

**Comment by keewis** [comment]
different default dtypes (e.g. `float32` instead of `float64`), I'd assume. The easiest fix would be to hard-code the dtypes, although that makes the code example a bit more verbose.

As for the formatting, is there a reason why the size is on the same line? Now that it is prefixed with `Size:`, we could just as well move it on a new line. Additionally, since this is the combined size (the sum of coordinates and data variables for datasets), maybe we should call it `Total size` or something similar? (to be clear, I'm hoping for a discussion, not requesting an immediate change)

**Comment by max-sixty** [comment]
> As for the formatting, is there a reason why the size is on the same line? Now that it is prefixed with `Size:`, we could just as well move it on a new line.

In general I'm a fan of more concise reprs, but no strong view. I agree the current version doesn't have perfect aesthetics.

> Additionally, since this is the combined size (the sum of coordinates and data variables for datasets), maybe we should call it `Total size` or something similar? (to be clear, I'm hoping for a discussion, not requesting an immediate change)

I figured that having it at the dataset level meant that it referred to the whole dataset. But again, no strong view!

**Comment by etienneschalk** [comment]
Hello,

### About windows failing tests

> The only issue I see is that the Windows tests seem to have different values for the size in the repr tests, [for example](https://github.com/pydata/xarray/actions/runs/7792432615/job/21250418463#step:9:1550):

Regarding the failing tests on windows, I had a similar issue recently, and a workaround is to use a "non-default" dtype that triggers the rendering of the dtype on the numpy array representation. 
However, to avoid these kind of tricky workarounds, the best would be to force numpy to show the dtype in the array's reprs, at least in a testing context to produce repeatable outputs. I made a quick search but could not find a way. Here is what I would like to do with numpy: forcing the printing of the dtype in the representation, something like:

```python
import numpy


np.set_printoptions(dtype="always")
```

### About conciseness of the `repr`

> In general I'm a fan of more concise reprs, but no strong view. I agree the current version doesn't have perfect aesthetics.

Regarding aesthetics: I agree that having a new line would look way more clean, and decluttering the header on the way

```
<xarray.Dataset>
Total size: 640B
...
```

However it would be at the cost of a newline, the question is if this newline acceptable, as it is an "uncompressible cost" (the rest of the repr changes until now only added info on existing lines). **Personnally, I would prefer the newline too, just to keep the aesthetics.**

Here is what it would looks like:

(A)

```
<xarray.Dataset>
Dimensions:  (x: 1)
Total size: 16B
Coordinates:
	y        (x) int64 8B dask.array<chunksize=(1,), meta=np.ndarray>
Dimensions without coordinates: x
Data variables:
	a        (x) int64 8B dask.array<chunksize=(1,), meta=np.ndarray>
```

(total size after dimensions, consistent with the inline repr) 

or

(B)


```
<xarray.Dataset>
Total size: 16B
Dimensions:  (x: 1)
Coordinates:
	y        (x) int64 8B dask.array<chunksize=(1,), meta=np.ndarray>
Dimensions without coordinates: x
Data variables:
	a        (x) int64 8B dask.array<chunksize=(1,), meta=np.ndarray>
```

(total size before dimensions, not consistent with the inline repr, but keep dims and coords grouped)

### About the definition of the size and total size of a DataArray

:warning: This may be out of scope of the required change, as in both Dataset and DataArray cases, the `nbytes` attribute is used. But this may seem surprising from a user perspective. (maybe for a future issue)

Here is an example of the output of a DataArray then Dataset repr with the proposed change:

```python
In [10]: xda = xr.DataArray([[1,2,3],[4,5,6]], coords = {"y": [40, 60], "x": [700, 800, 900]})

In [11]: xda
Out[11]: 
<xarray.DataArray (y: 2, x: 3)> Size: 48B
array([[1, 2, 3],
       [4, 5, 6]])
Coordinates:
  * y        (y) int64 16B 40 60
  * x        (x) int64 24B 700 800 900

In [12]: xr.Dataset({"var": xda})
Out[12]: 
<xarray.Dataset> Size: 88B
Dimensions:  (y: 2, x: 3)
Coordinates:
  * y        (y) int64 16B 40 60
  * x        (x) int64 24B 700 800 900
Data variables:
    var      (y, x) int64 48B 1 2 3 4 5 6
```

`48//6 == 8`,  this corresponds to the size of the data for the DataArray. However, the coordinates y and x also brings extra weight, that is not well represented here. For the Dataset however, containing the same DataArray, the `nbytes` also takes into account the coordinates: `16 + 24 + 48 == 88`

The doc indicates:

> xarray.DataArray.nbytes
> Total bytes consumed by the elements of this DataArray’s data.

> xarray.Dataset.nbytes
> Total bytes consumed by the data arrays of all variables in this dataset.

The distinction is not perfectly clear to me, indeed, a DataArray can also group other DataArrays (the coordinates)

```python
In [14]: xda.x
Out[14]: 
<xarray.DataArray 'x' (x: 3)> Size: 24B
array([700, 800, 900])
Coordinates:
  * x        (x) int64 24B 700 800 900
```

**Comment by max-sixty** [comment]
> Regarding the failing tests on windows, I had a similar issue recently, and a workaround is to use a "non-default" dtype that triggers the rendering of the dtype on the numpy array representation.

I think that's reasonable for the moment.

Tbc, it looks like it's not just the rendering — the actual dtype looks to be different, such that the size is correctly reported differently. So we need to make the dtypes be the same on windows. Does that make sense?

The tests are failing and should pass if this is done correctly.

**Comment by etienneschalk** [comment]
### About Windows 

> Tbc, it looks like it's not just the rendering — the actual dtype looks to be different, such that the size is correctly reported differently. So we need to make the dtypes be the same on windows. Does that make sense?

It seems Ubuntu + macOS defaults to 64 bits while Windows defaults on 32 bits. This default depends on the OS, I don't know if it's a "max" (eg if we could not use 64 bits on the Windows machines used by the CI because they have 32-bit OSes). I think it's worth trying to set the dtype to 64 bits at first, and see if it still fails. 

Indeed I looked for `dtype=np.int64` in the tests, and the occurences are tests that don't have any "skipif windows" decorator. However, the issue would move from the actual size to the `, dtype=int64` suffix in the numpy array repr.

Another option is to use the `ON_WINDOWS`  to make the 9 failing tests dependant on the OS. It is not the cleanest for sure

> I guess we could exclude those in windows if there isn't an easily reconcilable approach...

This is definitely the least effort, but excluding 9 tests for this PR seems overkill

This is the function used for numpy array representations. We can see the logic where it adds the suffix, there is no way to force print the dtype, or force not printing it.

https://github.com/numpy/numpy/blob/d35cd07ea997f033b2d89d349734c61f5de54b0d/numpy/core/arrayprint.py#L1487

What would solve this repeatable output issue would be to allow overriding this param:

```
def _array_repr_implementation(
        arr, max_line_width=None, precision=None, suppress_small=None,
+       skipdtype: bool | None = None,       
        array2string=array2string):
        ...
- skipdtype = dtype_is_implied(arr.dtype) and arr.size > 0
+ if skipdtype is None:
+     skipdtype = dtype_is_implied(arr.dtype) and arr.size > 0
```


### About the basicness of the current size printing algorithm 

Also, the current algorithm to print a human readable size is basic, and never shows decimal numbers. Maybe a better usage of the space could be made, eg:

```
Max used space: 5 letters
999kB 

What could be improved is:

9kB 
vvvvv
9.9kB  (use 5 letter space)
```

Indeed, the current size is not reliable and is just an estimation that should not replace the integer value returned by `nbytes` :thinking:

**Comment by max-sixty** [comment]
I don't really understand why the tests fail. It seems that on windows, 3 int64 values take up 48 bytes of space??

https://github.com/pydata/xarray/actions/runs/7792432615/job/21250418347?pr=8702#step:9:419

```python
 self = <xarray.tests.test_dataarray.TestDataArray object at 0x00000234F4DB19D0>

    def test_repr(self) -> None:
        v = Variable(["time", "x"], [[1, 2, 3], [4, 5, 6]], {"foo": "bar"})
        coords = {"x": np.arange(3, dtype=np.int64), "other": np.int64(0)}
        data_array = DataArray(v, coords, name="my_variable")
        expected = dedent(
            """\
            <xarray.DataArray 'my_variable' (time: 2, x: 3)> Size: 48B
            array([[1, 2, 3],
                   [4, 5, 6]])
            Coordinates:
              * x        (x) int64 24B 0 1 2
                other    int64 8B 0
            Dimensions without coordinates: time
            Attributes:
                foo:      bar"""
        )
>       assert expected == repr(data_array)
E       AssertionError: assert '<xarray.Data...foo:      bar' == '<xarray.Data...foo:      bar'
E         
E         Skipping 45 identical leading characters in diff, use -v to show
E         Skipping 166 identical trailing characters in diff, use -v to show
E         - 3)> Size: 24B
E         ?           -
E         + 3)> Size: 48B
E         ?            +
E           array([
```

Unless anyone has a better idea, I think skipping on Windows is OK.

---

> Also, the current algorithm to print a human readable size is basic, and never shows decimal numbers. Maybe a better usage of the space could be made, eg:

I would keep it simple, and not conditionally change precision, at least for the moment.

**Comment by etienneschalk** [comment]
> Unless anyone has a better idea, I think skipping on Windows is OK.

https://github.com/pydata/xarray/pull/8702/files/e98a97d3085e7dc3b1bcb11ac8af012fc1acc1c4..e2db82a6d2a322e6ed18ebb0cb2bd696458b540c

For big tests, I used a `skipif` approach to test both Windows and non-Windows. For smaller, I added the condition in-test. 

It adds many `ON_WINDOWS` constants, but is unavoidable, as the representation including size is OS-dependant in the current CI. (It seems `win32` is not reliable to determine if we are on 32 bit OS as all windows would return `win32`.

Crazy how a simple "add size to repr" issue turned out to "platform-dependant shenanigans" ! 

Links I consulted:

- https://stackoverflow.com/questions/2144748/is-it-safe-to-use-sys-platform-win32-check-on-64-bit-python
- https://stackoverflow.com/questions/36278590/numpy-array-dtype-is-coming-as-int32-by-default-in-a-windows-10-64-bit-machine


> I would keep it simple, and not conditionally change precision, at least for the moment.

OK!

**Comment by keewis** [comment]
> I don't really understand why the tests fail. It seems that on windows, 3 int64 values take up 48 bytes of space??

`int64` has 8 bytes per element, so I agree, 3 values should be 24 bytes. However, if you look at the data it's actually a 2×3 array, and with 6 values 48 bytes makes sense. And looking at the traceback, the issue is that on windows the data only has a size of 24 bytes, which means that it is using `int32` as a dtype, with 4 bytes per element. Which tells me that once again the default size is the issue.

**Comment by max-sixty** [comment]
> `int64` has 8 bytes per element, so I agree, 3 values should be 24 bytes. However, if you look at the data it's actually a 2×3 array, and with 6 values 48 bytes makes sense. And looking at the traceback, the issue is that on windows the data only has a size of 24 bytes, which means that it is using `int32` as a dtype, with 4 bytes per element. Which tells me that once again the default size is the issue.

Ah, very good point. In the case above it's that this array:

```
        v = Variable(["time", "x"], [[1, 2, 3], [4, 5, 6]], {"foo": "bar"})
```

is being cast to the default size. So we can instead cast it with `.astype(int64)` or similar (or create from `np.arange`, whatever is easier.

**Comment by etienneschalk** [comment]
Tests pass with a differenciation between windows and non-windows env. 

**I tagged `test_dask_roundtrip` as flaky (xfail) as it frequently failed in my previous CI runs.** 

Hopefully this is acceptable! 

The current state of the PR don't integrate recent discussions about total size. The solution implemented is the one described from https://github.com/pydata/xarray/issues/8690#issuecomment-1926187361 . Maybe further discussions should happen on the original issue rather than this PR

**Comment by max-sixty** [comment]
I would be +1 on merging.

I think the windows issues could be better handled by setting the dtype (i.e. https://github.com/pydata/xarray/pull/8702#issuecomment-1930729027, building on @keewis 's observation), but we can also do that in another PR, and this has a large enough blast radius that it would be better to merge sooner.

Would someone else agree? (Or feel free to just hit the button...)

**Review by max-sixty — APPROVED** [review]

**Comment by etienneschalk** [comment]
@max-sixty 

By setting the dtype, I think that only a part of the issue would be solved as numpy would print out `, dtype=np.int64`, as numpy always prints out non-default dtypes. This behaviour does not seem to be able to being turned off unfortunately.  

So the repr would still be different and the need to differentiate between Windows and non-Windows environments still remains :thinking: 

This comment on the numpy repo is interesting: https://github.com/numpy/numpy/issues/9464#issuecomment-374476377

**Comment by max-sixty** [comment]
> By setting the dtype, I think that only a part of the issue would be solved as numpy would print out `, dtype=np.int64`, as numpy always prints out non-default dtypes. This behaviour does not seem to be able to being turned off unfortunately.

What's an xarray object that is explicitly typed that would show different reprs in linux/mac vs windows?

**Comment by max-sixty** [comment]
@pydata/xarray could someone second the approval here?

**Review by dcherian — APPROVED** [review]

**Comment by max-sixty** [comment]
@etienneschalk great work!

Thanks also for the issues into pytest-accept.

I wanted to merge asap so we didn't get merge conflicts. If you're up for simplifying the formatting tests — assuming I'm not mistaken above — that would be a very nice 2nd PR...

**Comment by etienneschalk** [comment]
Thanks @max-sixty!

> > By setting the dtype, I think that only a part of the issue would be solved as numpy would print out `, dtype=np.int64`, as numpy always prints out non-default dtypes. This behaviour does not seem to be able to being turned off unfortunately.
> 
> What's an xarray object that is explicitly typed that would show different reprs in linux/mac vs windows?

I detailed the issue in a previous comment https://github.com/pydata/xarray/pull/8702#issuecomment-1930614831 

If we use `dtype=np.int32`, macOS and Linux will add a `, dtype=np.int32` in the array repr and if we use `dtype=np.int64`, then Windows will add a `, dtype=np.int64` in the array repr. 

Also, even when providing explitly a dtype, if the dtype is a default, it won't show up in the repr. Here is an example on my machine (Linux):

_numpy-only_

```python
In [3]: import numpy as np

In [4]: np.array([1,2,3])
Out[4]: array([1, 2, 3])

In [5]: np.array([1,2,3], dtype=np.int64)
Out[5]: array([1, 2, 3])

In [6]: np.array([1,2,3], dtype=np.int32)
Out[6]: array([1, 2, 3], dtype=int32)
```

_xarray (delegating repr to numpy)_

```python
In [14]: import xarray as xr

In [15]: xr.DataArray(np.array([1,2,3]))
Out[15]: 
<xarray.DataArray (dim_0: 3)>
array([1, 2, 3])
Dimensions without coordinates: dim_0

In [16]: xr.DataArray(np.array([1,2,3], dtype=np.int64))
Out[16]: 
<xarray.DataArray (dim_0: 3)>
array([1, 2, 3])
Dimensions without coordinates: dim_0

In [17]: xr.DataArray(np.array([1,2,3], dtype=np.int32))
Out[17]: 
<xarray.DataArray (dim_0: 3)>
array([1, 2, 3], dtype=int32)
Dimensions without coordinates: dim_0
```

I expected the opposite on the Windows CI ; `, dtype=int64` to be shown. So the only way to always get an explicit dtype repr would be to: 
- either, force its display (which is not possible, see https://github.com/numpy/numpy/blob/d35cd07ea997f033b2d89d349734c61f5de54b0d/numpy/core/arrayprint.py#L1487 ; the `skipdtype` is not overridable with a kwarg, and experiment with changing the format [printoptions](https://numpy.org/doc/stable/reference/generated/numpy.set_printoptions.html) to legacy was not concluant)
- or, find a dtype that is a non-default on all of the 3 CI platforms (eg `int16`)

> I wanted to merge asap so we didn't get merge conflicts. If you're up for simplifying the formatting tests — assuming I'm not mistaken above — that would be a very nice 2nd PR...

I still need to confirm all I said above by more experimentation. Actually I can add such "repr testing" in a next PR yes, to try to make a catalog of all these problematic reprs

**Comment by max-sixty** [comment]
@etienneschalk sorry, you're completely correct. I was thinking about the Dataset repr. But you're correct that the DataArray repr just inherits from numpy.

So I don't have any strong views about better ways of doing this — the existing way seems OK. Another approach would be to just `re.sub` `Size: \d+B` in all the reprs...

**Comment by djhoese** [comment]
Is there a reason the title of this PR can't be updated to reflect what the description was edited/updated to say: this is not opt-in.

**Pull request merged**